phd-tests/framework/src/disk/fat.rs

```diff
@@ -0,0 +1,249 @@
+/// - 4085 <= clusters <= 65524: FAT16
+/// - clusters >= 65525: FAT32
+///
+/// There is a small catch-22 here: a volume's format depends on the number of
+/// clusters it has, but the number of clusters in the volume depends on the
+/// size of the filesystem metadata, which depends on the volume's format.
+/// This module avoids the problem by requiring that the total volume size is
+/// less than or equal to 4,084 sectors. The number of clusters will never be
+/// greater than this, so the filesystem will always be FAT12, which makes it
+/// possible to compute its overhead size.
+fn overhead_sectors() -> Sectors {
+    let dir_entry_bytes = ROOT_DIRECTORY_ENTRIES * BYTES_PER_DIRECTORY_ENTRY;
+    let dir_entry_sectors = Sectors::needed_for_bytes(dir_entry_bytes);
+
+    // To compute the size of the FAT, conservatively assume that all the
+    // sectors on the disk are in addressable clusters, figure out how many
+    // bytes that would take, and convert to sectors. (In practice, some of the
+    // sectors are used for overhead and won't have entries in the FAT, but this
+    // gives an upper bound.)
+    let max_sectors = Sectors::needed_for_bytes(MAX_DISK_SIZE_BYTES);
+
+    // FAT12 tables use 12 bits per cluster entry.
+    let cluster_bits = max_sectors.0 * 12;
+    let cluster_bytes = cluster_bits.div_ceil(8);
+    let sectors_per_table = Sectors::needed_for_bytes(cluster_bytes);
+
+    // The total overhead size is one sector (for the BPB) plus the sectors
+    // needed for regions 2 and 3 (the tables themselves and the root directory
+    // entries). Note that there are multiple tables per volume!
+    Sectors(1) + (sectors_per_table * TABLES_PER_VOLUME) + dir_entry_sectors
+}
+
+/// Yields the number of sectors in this module's FAT volumes that can be used
+/// by files.
+fn total_usable_sectors() -> Sectors {
+    Sectors::needed_for_bytes(MAX_DISK_SIZE_BYTES) - overhead_sectors()
+}
+
+#[derive(Clone, Debug)]
+struct File {
+    path: String,
+    contents: Vec<u8>,
+}
+
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error(
+        "insufficient space for new file: {0} sectors required, {1} available"
+    )]
+    NoSpace(usize, usize),
+}
+
+/// Builds a collection of files that can be extruded as an array of bytes
+/// containing a FAT filesystem with the collected files.
+#[derive(Clone, Default, Debug)]
+pub struct FatFilesystem {
+    files: Vec<File>,
+    sectors_remaining: Sectors,
+}
+
+impl FatFilesystem {
+    /// Creates a new file collection.
+    pub fn new() -> Self {
+        Self { files: vec![], sectors_remaining: total_usable_sectors() }
+    }
+
+    /// Adds a file with the supplied `contents` that will appear in the
+    /// resulting file system in the supplied `path`.
+    ///
+    /// N.B. `path` is interpreted relative to the root of the file system and
+    ///      should not have a leading '/'.
+    pub fn add_file_from_str(
+        &mut self,
+        path: &str,
```

Comment:
The `fatfs` crate speaks VFAT, so it can handle long filenames containing UTF-16 characters. (There are still some restrictions--e.g. it will reject an attempt to create a file if it can't find an unused MS-DOS short name for it--but those errors will get bubbled up when a caller tries to create a disk that runs afoul of these rules.)

There is one class of error we can catch up front, though (and it's one that makes me disinclined to make this a `Path`): if the supplied path contains a separator, `fatfs` will (correctly) try to traverse the filesystem's current directory tree and put the file in the requested directory. That will never work in PHD because the `FatFilesystem` struct doesn't (currently) let users create directories. So I've tweaked this in 3e77600 to make it clearer that this needs to be a bare filename (and to bail on adding a file if the name contains a '/' character).